apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/DcGeoRDALoadJob.java

```diff
@@ -0,0 +1,149 @@
+  public static final String FAILURES_METER_NAME =
+      MetricRegistry.name(DcGeoRDALoadJob.class.getSimpleName(), "failures");
+  public static final String SUCCESSES_METER_NAME =
+      MetricRegistry.name(DcGeoRDALoadJob.class.getSimpleName(), "successes");
+  public static final String PROCESSED_METER_NAME =
+      MetricRegistry.name(DcGeoRDALoadJob.class.getSimpleName(), "processed");
+
+  private final Config config;
+  private final Callable<RDASource<PreAdjudicatedClaim>> sourceFactory;
+  private final Callable<RDASink<PreAdjudicatedClaim>> sinkFactory;
+  private final Meter callsMeter;
+  private final Meter failuresMeter;
+  private final Meter successesMeter;
+  private final Meter processedMeter;
+
+  public DcGeoRDALoadJob(
+      Config config,
+      Callable<RDASource<PreAdjudicatedClaim>> sourceFactory,
+      Callable<RDASink<PreAdjudicatedClaim>> sinkFactory,
+      MetricRegistry appMetrics) {
+    this.config = config;
+    this.sourceFactory = sourceFactory;
+    this.sinkFactory = sinkFactory;
+    callsMeter = appMetrics.meter(CALLS_METER_NAME);
+    failuresMeter = appMetrics.meter(FAILURES_METER_NAME);
+    successesMeter = appMetrics.meter(SUCCESSES_METER_NAME);
+    processedMeter = appMetrics.meter(PROCESSED_METER_NAME);
+  }
+
+  /**
+   * Factory method to construct a new job instance using standard parameters.
+   *
+   * @param appMetrics MetricRegistry used to track operational metrics
+   * @return a DcGeoRDALoadJob instance suitable for use by PipelineManager.
+   */
+  public static DcGeoRDALoadJob newDcGeoRDALoadJob(MetricRegistry appMetrics) {
+    return new DcGeoRDALoadJob(
+        new Config(),
+        () -> new SkeletonRDASource(appMetrics),
+        () -> new SkeletonRDASink(appMetrics),
+        appMetrics);
+  }
+
+  public Duration getScanInterval() {
+    return config.getScanInterval();
+  }
+
+  @Override
+  public PipelineJobOutcome call() throws Exception {
+    final long startMillis = System.currentTimeMillis();
+    int processedCount = 0;
+    Exception error = null;
+    try {
+      callsMeter.mark();
+      try (RDASource<PreAdjudicatedClaim> source = sourceFactory.call();
+          RDASink<PreAdjudicatedClaim> sink = sinkFactory.call()) {
+        processedCount =
+            source.retrieveAndProcessObjects(
+                config.getMaxObjectsPerCall(), config.getBatchSize(), config.getMaxRunTime(), sink);
+      }
+    } catch (ProcessingException ex) {
+      processedCount += ex.getProcessedCount();
+      error = ex.getCause();
+    } catch (Exception ex) {
+      error = ex;
+    }
+    processedMeter.mark(processedCount);
+    final long stopMillis = System.currentTimeMillis();
+    LOGGER.info("processed {} objects in {} ms", processedCount, stopMillis - startMillis);
+    if (error != null) {
+      failuresMeter.mark();
+      LOGGER.error("processing aborted by an exception: message={}", error.getMessage(), error);
+      throw error;
```

Comment:
I always recommend strongly against rethrows like this, as it discards the stack from the actual cause, which will make debugging very challenging. Instead, use a wrap-and-rethrow, which will preserve and extend the original stack trace.

Reply:
The stack trace is established when an exception is created, not when it's thrown or rethrown.  That can be verified with a unit test.  Rethrowing preserves the original stack trace.

If you'd prefer the exception be wrapped I can certainly add that.  I just find nested exceptions can add excessive bulk to the logs when they aren't actuall required.

Reply:
That is WILD: I've had that wrong in my head for, like, a damn decade. Thanks for the correction!

I do still think it's important to always wrap-and-rethrow, though, because otherwise your stack trace is painting a somewhat misleading picture about the control flow.